router/proxy/reverseproxy.go

```diff
@@ -0,0 +1,332 @@
+	if err != nil {
+		p.logf("router: hijack failed: %v", err)
+		return
+	}
+	joinConns(uconn, &streamConn{bufrw.Reader, dconn})
+}
+
+func (p *ReverseProxy) writeResponse(rw http.ResponseWriter, res *http.Response) {
+	// remove hop-by-hop headers as specified in connection options.
+	for _, h := range hopHeaders {
+		res.Header.Del(h)
+	}
+
+	// remove the Upgrade header and headers referenced in the Connection
+	// header if HTTP < 1.1 or if Connection header didn't contain "upgrade":
+	// https://tools.ietf.org/html/rfc7230#section-6.7
+	if !res.ProtoAtLeast(1, 1) || !isConnectionUpgrade(res.Header) {
+		res.Header.Del("Upgrade")
+
+		// A proxy or gateway MUST parse a received Connection header field before a
+		// message is forwarded and, for each connection-option in this field, remove
+		// any header field(s) from the message with the same name as the
+		// connection-option, and then remove the Connection header field itself (or
+		// replace it with the intermediary's own connection options for the
+		// forwarded message): https://tools.ietf.org/html/rfc7230#section-6.1
+		tokens := strings.Split(res.Header.Get("Connection"), ",")
+		for _, hdr := range tokens {
+			res.Header.Del(hdr)
+		}
+		res.Header.Del("Connection")
+	}
+
+	copyHeader(rw.Header(), res.Header)
+
+	rw.WriteHeader(res.StatusCode)
+	p.copyResponse(rw, res.Body)
+}
+
+func isConnectionUpgrade(h http.Header) bool {
+	for _, token := range strings.Split(h.Get("Connection"), ",") {
+		if v := strings.ToLower(strings.TrimSpace(token)); v == "upgrade" {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *ReverseProxy) copyResponse(dst io.Writer, src io.Reader) {
+	if p.FlushInterval != 0 {
+		if wf, ok := dst.(writeFlusher); ok {
+			mlw := &maxLatencyWriter{
+				dst:     wf,
+				latency: p.FlushInterval,
+				done:    make(chan bool),
+			}
+			go mlw.flushLoop()
+			defer mlw.stop()
+			dst = mlw
+		}
+	}
+
+	io.Copy(dst, src)
+}
+
+func (p *ReverseProxy) logf(format string, args ...interface{}) {
+	if p.ErrorLog != nil {
+		p.ErrorLog.Printf(format, args...)
+	} else {
+		log.Printf(format, args...)
+	}
+}
+
+func copyHeader(dst, src http.Header) {
+	for k, vv := range src {
+		for _, v := range vv {
+			dst.Add(k, v)
+		}
+	}
+}
+
+type closeWriter interface {
+	CloseWrite() error
+}
+
+func closeWrite(conn net.Conn) {
+	if cw, ok := conn.(closeWriter); ok {
+		cw.CloseWrite()
+	} else {
+		conn.Close()
+	}
+}
+
+func joinConns(uconn, dconn net.Conn) {
+	done := make(chan struct{})
+
+	go func() {
+		io.Copy(uconn, dconn)
+		closeWrite(uconn)
+		done <- struct{}{}
+	}()
+
+	io.Copy(dconn, uconn)
+	closeWrite(dconn)
+	<-done
+}
+
+func prepareRequest(req *http.Request) *http.Request {
+	outreq := new(http.Request)
+	*outreq = *req // includes shallow copies of maps, but okay
+
+	// Pass the Request-URI verbatim without any modifications
+	outreq.URL.Opaque = strings.Split(strings.TrimPrefix(req.RequestURI, req.URL.Scheme+":"), "?")[0]
```

Comment:
There is already a test for the request URI (assuming this is what you mean): https://github.com/flynn/flynn/blob/2a0f1e7ac1f95fdd5585dd62e659210378f918c8/router/http_test.go#L921-L946

I've also just opened #876 with a test to guarantee query params.